uber/automated_emails.py

```diff
@@ -46,46 +64,235 @@ def computed_subject(self, x):
+        return render('emails/' + self.template, dict({model: model_instance}, **self.extra_data))
 
-    def render(self, x):
-        model = getattr(x, 'email_model_name', x.__class__.__name__.lower())
-        return render('emails/' + self.template, dict({model: x}, **self.extra_data))
+    def really_send(self, model_instance):
+        """
+        Actually send an email to a particular model instance (i.e. a particular attendee).
+
+        Doesn't perform any kind of checks at all if we should be sending this, just immediately sends the email
+        no matter what.
 
-    def send(self, x, raise_errors=True):
+        NOTE: use send_if_possible() instead of calling this method unless you 100% know what you're doing.
+        """
         try:
             subject = self.computed_subject(x)
             format = 'text' if self.template.endswith('.txt') else 'html'
-            send_email(self.sender, x.email, subject, self.render(x), format, model=x, cc=self.cc, ident=self.ident)
+            send_email(self.sender, model_instance.email, subject,
+                       self.render(model_instance), format,
+                       model=model_instance, cc=self.cc, ident=self.ident)
         except:
-            log.error('error sending {!r} email to {}', self.subject, x.email, exc_info=True)
-            if raise_errors:
-                raise
+            log.error('error sending {!r} email to {}', self.subject, model_instance.email, exc_info=True)
+            raise
+
+    @property
+    def when_txt(self):
+        """
+        Return a textual description of when the date filters are active for this email category.
+        """
+
+        return '\n'.join([filter.active_when for filter in self.when])
+
+
+class SendAllAutomatedEmailsJob:
+
+    # save information about the last time the daemon ran so that we can display stats on things like
+    # unapproved emails/etc
+    last_result = dict()
+
+    run_lock = threading.Lock()
+
+    @classmethod
+    def send_all_emails(cls, raise_errors=False):
+        """ Helper method to start a run of our automated email processing """
+        cls().run(raise_errors)
+
+    @timed
+    def run(self, raise_errors=False):
+        """
+        Do one run of our automated email service.  Call this periodically to send any emails that should go out
+        automatically.
+
+        This will NOT run if we're on-site, or not configured to send emails.
+
+        :param raise_errors: If False, exceptions are squashed during email sending and we'll try the next email.
+        """
+        allowed_to_run = not c.AT_THE_CON and (c.DEV_BOX or c.SEND_EMAILS)
+        if not allowed_to_run:
+            return
+
+        if not self.run_lock.acquire(blocking=False):
+            log.warn("can't acquire lock for email daemon, skipping run.")
+            return
+
+        self._run(raise_errors)
+
+        self.run_lock.release()
```

Comment:
yea, on looking at it again, agreed.  It took a couple iterations of refactoring to get it into its current form, and I thnk the right call is the try/except like you said, and removing @swallow_exceptions